// eas-cli/build/credentials/manager/SelectPlatform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectPlatform = void 0;
const platform_1 = require("../../platform");
const ManageAndroid_1 = require("./ManageAndroid");
const ManageIos_1 = require("./ManageIos");
class SelectPlatform {
    constructor(actor, graphqlClient, analytics, projectInfo, getDynamicPrivateProjectConfigAsync, flagPlatform) {
        this.actor = actor;
        this.graphqlClient = graphqlClient;
        this.analytics = analytics;
        this.projectInfo = projectInfo;
        this.getDynamicPrivateProjectConfigAsync = getDynamicPrivateProjectConfigAsync;
        this.flagPlatform = flagPlatform;
    }
    async runAsync() {
        const platform = await (0, platform_1.selectPlatformAsync)(this.flagPlatform);
        if (platform === 'ios') {
            return await new ManageIos_1.ManageIos(this, process.cwd()).runAsync();
        }
        return await new ManageAndroid_1.ManageAndroid(this, process.cwd()).runAsync();
    }
}
exports.SelectPlatform = SelectPlatform;
