// @sentry/react-native/dist/js/utils/environment.js
import { version as RNV } from 'react-native/Libraries/Core/ReactNativeVersion';
import { RN_GLOBAL_OBJ } from '../utils/worldwide';
/** Checks if the React Native Hermes engine is running */
export function isHermesEnabled() {
    return !!RN_GLOBAL_OBJ.HermesInternal;
}
/** Checks if the React Native TurboModules are enabled */
export function isTurboModuleEnabled() {
    return RN_GLOBAL_OBJ.__turboModuleProxy != null;
}
/** Checks if the React Native Fabric renderer is running */
export function isFabricEnabled() {
    return RN_GLOBAL_OBJ.nativeFabricUIManager != null;
}
/** Returns React Native Version as semver string */
export function getReactNativeVersion() {
    return `${RNV.major}.${RNV.minor}.${RNV.patch}${RNV.prerelease != null ? `-${RNV.prerelease}` : ''}`;
}
/** Checks if Expo is present in the runtime */
export function isExpo() {
    return RN_GLOBAL_OBJ.expo != null;
}
/** Returns Hermes Version if hermes is present in the runtime */
export function getHermesVersion() {
    return (RN_GLOBAL_OBJ.HermesInternal &&
        RN_GLOBAL_OBJ.HermesInternal.getRuntimeProperties &&
        RN_GLOBAL_OBJ.HermesInternal.getRuntimeProperties()['OSS Release Version']);
}
