// @sentry/react-native/dist/js/integrations/reactnativeinfo.js
import { __awaiter } from "tslib";
import { getHermesVersion, getReactNativeVersion, isExpo, isFabricEnabled, isHermesEnabled, isTurboModuleEnabled, } from '../utils/environment';
/** Loads React Native context at runtime */
export class ReactNativeInfo {
    constructor() {
        /**
         * @inheritDoc
         */
        this.name = ReactNativeInfo.id;
    }
    /**
     * @inheritDoc
     */
    setupOnce(addGlobalEventProcessor) {
        addGlobalEventProcessor((event, hint) => __awaiter(this, void 0, void 0, function* () {
            const reactNativeError = (hint === null || hint === void 0 ? void 0 : hint.originalException) ? hint === null || hint === void 0 ? void 0 : hint.originalException : undefined;
            const reactNativeContext = {
                turbo_module: isTurboModuleEnabled(),
                fabric: isFabricEnabled(),
                react_native_version: getReactNativeVersion(),
                expo: isExpo(),
            };
            if (isHermesEnabled()) {
                reactNativeContext.js_engine = 'hermes';
                const hermesVersion = getHermesVersion();
                if (hermesVersion) {
                    reactNativeContext.hermes_version = getHermesVersion();
                }
            }
            else if (reactNativeError === null || reactNativeError === void 0 ? void 0 : reactNativeError.jsEngine) {
                reactNativeContext.js_engine = reactNativeError.jsEngine;
            }
            if (reactNativeContext.js_engine === 'hermes') {
                event.tags = Object.assign({ hermes: 'true' }, event.tags);
            }
            if (reactNativeError === null || reactNativeError === void 0 ? void 0 : reactNativeError.componentStack) {
                reactNativeContext.component_stack = reactNativeError.componentStack;
            }
            event.contexts = Object.assign({ react_native_context: reactNativeContext }, event.contexts);
            return event;
        }));
    }
}
/**
 * @inheritDoc
 */
ReactNativeInfo.id = 'ReactNativeInfo';
