// @expo/eas-build-job/dist/common.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheSchema = exports.EnvironmentSecretsSchema = exports.EnvironmentSecretType = exports.EnvSchema = exports.ArchiveSourceSchema = exports.BuildTrigger = exports.ArchiveSourceType = exports.Platform = exports.Workflow = exports.BuildMode = void 0;
const joi_1 = __importDefault(require("joi"));
var BuildMode;
(function (BuildMode) {
    BuildMode["BUILD"] = "build";
    BuildMode["RESIGN"] = "resign";
    BuildMode["CUSTOM"] = "custom";
})(BuildMode = exports.BuildMode || (exports.BuildMode = {}));
var Workflow;
(function (Workflow) {
    Workflow["GENERIC"] = "generic";
    Workflow["MANAGED"] = "managed";
    Workflow["UNKNOWN"] = "unknown";
})(Workflow = exports.Workflow || (exports.Workflow = {}));
var Platform;
(function (Platform) {
    Platform["ANDROID"] = "android";
    Platform["IOS"] = "ios";
})(Platform = exports.Platform || (exports.Platform = {}));
var ArchiveSourceType;
(function (ArchiveSourceType) {
    ArchiveSourceType["NONE"] = "NONE";
    ArchiveSourceType["S3"] = "S3";
    ArchiveSourceType["URL"] = "URL";
    ArchiveSourceType["PATH"] = "PATH";
    ArchiveSourceType["GCS"] = "GCS";
    ArchiveSourceType["GIT"] = "GIT";
})(ArchiveSourceType = exports.ArchiveSourceType || (exports.ArchiveSourceType = {}));
var BuildTrigger;
(function (BuildTrigger) {
    BuildTrigger["EAS_CLI"] = "EAS_CLI";
    BuildTrigger["GIT_BASED_INTEGRATION"] = "GIT_BASED_INTEGRATION";
})(BuildTrigger = exports.BuildTrigger || (exports.BuildTrigger = {}));
exports.ArchiveSourceSchema = joi_1.default.object({
    type: joi_1.default.string()
        .valid(...Object.values(ArchiveSourceType))
        .required(),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.GCS }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.GCS).required(),
        bucketKey: joi_1.default.string().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.S3 }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.S3).required(),
        bucketKey: joi_1.default.string().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.URL }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.URL).required(),
        url: joi_1.default.string().uri().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.GIT }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.GIT).required(),
        repositoryUrl: joi_1.default.string().required(),
        gitRef: joi_1.default.string().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.PATH }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.PATH).required(),
        path: joi_1.default.string().required(),
    }),
});
exports.EnvSchema = joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.string());
var EnvironmentSecretType;
(function (EnvironmentSecretType) {
    EnvironmentSecretType["STRING"] = "string";
    EnvironmentSecretType["FILE"] = "file";
})(EnvironmentSecretType = exports.EnvironmentSecretType || (exports.EnvironmentSecretType = {}));
exports.EnvironmentSecretsSchema = joi_1.default.array().items(joi_1.default.object({
    name: joi_1.default.string().required(),
    value: joi_1.default.string().required(),
    type: joi_1.default.string()
        .valid(...Object.values(EnvironmentSecretType))
        .required(),
}));
exports.CacheSchema = joi_1.default.object({
    disabled: joi_1.default.boolean().default(false),
    clear: joi_1.default.boolean().default(false),
    key: joi_1.default.string().allow('').max(128),
    cacheDefaultPaths: joi_1.default.boolean(),
    customPaths: joi_1.default.array().items(joi_1.default.string()),
    paths: joi_1.default.array().items(joi_1.default.string()).default([]),
});
