// @expo/steps/dist_commonjs/BuildTemporaryFiles.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cleanUpStepTemporaryDirectoriesAsync = exports.createTemporaryOutputsDirectoryAsync = exports.saveScriptToTemporaryFileAsync = void 0;
const path_1 = __importDefault(require("path"));
const promises_1 = __importDefault(require("fs/promises"));
const uuid_1 = require("uuid");
async function saveScriptToTemporaryFileAsync(ctx, stepId, scriptContents) {
    const scriptsDir = getTemporaryScriptsDirPath(ctx, stepId);
    await promises_1.default.mkdir(scriptsDir, { recursive: true });
    const temporaryScriptPath = path_1.default.join(scriptsDir, `${(0, uuid_1.v4)()}.sh`);
    await promises_1.default.writeFile(temporaryScriptPath, scriptContents);
    return temporaryScriptPath;
}
exports.saveScriptToTemporaryFileAsync = saveScriptToTemporaryFileAsync;
async function createTemporaryOutputsDirectoryAsync(ctx, stepId) {
    const directory = getTemporaryOutputsDirPath(ctx, stepId);
    await promises_1.default.mkdir(directory, { recursive: true });
    return directory;
}
exports.createTemporaryOutputsDirectoryAsync = createTemporaryOutputsDirectoryAsync;
async function cleanUpStepTemporaryDirectoriesAsync(ctx, stepId) {
    if (ctx.skipCleanup) {
        return;
    }
    const stepTemporaryDirectory = getTemporaryStepDirPath(ctx, stepId);
    await promises_1.default.rm(stepTemporaryDirectory, { recursive: true, force: true });
    ctx.logger.debug({ stepTemporaryDirectory }, 'Removed step temporary directory');
}
exports.cleanUpStepTemporaryDirectoriesAsync = cleanUpStepTemporaryDirectoriesAsync;
function getTemporaryStepDirPath(ctx, stepId) {
    return path_1.default.join(ctx.baseWorkingDirectory, 'steps', stepId);
}
function getTemporaryScriptsDirPath(ctx, stepId) {
    return path_1.default.join(getTemporaryStepDirPath(ctx, stepId), 'scripts');
}
function getTemporaryOutputsDirPath(ctx, stepId) {
    return path_1.default.join(getTemporaryStepDirPath(ctx, stepId), 'outputs');
}
